File: Example/QDNavigationBar/QDConfigEditor/QDConfigSwitchView.m
//
//  QDConfigSwitchView.m
//  QDCustomNavigation_Example
//
//  Created by Sinno on 2020/8/5.
//  Copyright © 2020 sinno93. All rights reserved.
//

#import "QDConfigSwitchView.h"
#import <Masonry/Masonry.h>

@interface QDConfigSwitchView ()
@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UISwitch *switchView;
@end

@implementation QDConfigSwitchView

#pragma mark - Initialization
// 初始化方法
- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        [self configSubviews];
    }
    return self;
}

#pragma mark - Super methods

#pragma mark - Action Methods
- (void)onValueChanged {
    if (self.valueChanged) {
        self.valueChanged(self.switchView.on);
    }
}

#pragma mark - Public Methods
// 公开的方法
- (void)setTitle:(NSString *)title {
    self.titleLabel.text = title;
}

- (NSString *)title {
    return self.titleLabel.text;
}

- (void)setOn:(BOOL)on {
    self.switchView.on = on;
}

- (BOOL)on {
    return self.switchView.on;
}

#pragma mark - Private Methods
// 私有的方法
- (void)configSubviews {
    [self addSubview:self.titleLabel];
    [self addSubview:self.switchView];

    [self.titleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(self);
        make.leading.equalTo(self).offset(15);
    }];
    
    [self.switchView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(self);
        make.trailing.equalTo(self).offset(-25);
    }];
    
}
#pragma mark - Getter && Setter
// Getter和Setter方法
- (UILabel *)titleLabel {
    if (!_titleLabel) {
        UILabel *label = [[UILabel alloc] init];
        label.font = [UIFont systemFontOfSize:20.0];
        label.textColor = UIColor.blackColor;
        _titleLabel = label;
    }
    return _titleLabel;
}

- (UISwitch *)switchView {
    if (!_switchView) {
        _switchView = [[UISwitch alloc] init];
        [_switchView addTarget:self action:@selector(onValueChanged) forControlEvents:UIControlEventValueChanged];
    }
    return _switchView;
}

@end

File: Example/QDNavigationBar/QDConfigEditor/QDConfigSwitchView.swift
//
//  QDConfigSwitchView.swift
//  QDNavigationBar_Example
//
//  Created by zqd on 2024/6/5.
//  Copyright © 2024 CocoaPods. All rights reserved.
//

import UIKit
import SnapKit

class QDConfigSwitchView: UIView {
    
    private var titleLabel: UILabel = {
        let label = UILabel()
        label.font = UIFont.systemFont(ofSize: 20.0)
        label.textColor = .black
        return label
    }()
    
    private var switchView: UISwitch = {
        let switchView = UISwitch()
        return switchView
    }()
    
    var valueChanged: ((Bool) -> Void)?
    
    var title: String? {
        get {
            return titleLabel.text
        }
        set {
            titleLabel.text = newValue
        }
    }
    
    var on: Bool {
        get {
            return switchView.isOn
        }
        set {
            switchView.isOn = newValue
        }
    }
    
    // MARK: - Initialization
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        configSubviews()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        configSubviews()
    }
    
    // MARK: - Private Methods
    
    private func configSubviews() {
        addSubview(titleLabel)
        addSubview(switchView)
        
        titleLabel.snp.makeConstraints { make in
            make.centerY.equalToSuperview()
            make.leading.equalToSuperview().offset(15)
        }
        
        switchView.snp.makeConstraints { make in
            make.centerY.equalToSuperview()
            make.trailing.equalToSuperview().offset(-25)
        }
        
        switchView.addTarget(self, action: #selector(onValueChanged), for: .valueChanged)
    }
    
    @objc private func onValueChanged() {
        valueChanged?(switchView.isOn)
    }
}

